### DISCUSSION TITLE
Error in post-command-hook (corfu--post-command): (wrong-type-argument number-or-marker-p nil)

### DISCUSSION THREAD

**Discussion by aaronjensen** [minad/corfu#470] [Q&A] [answered]
I'm seeing an error when I hit `tab` when I'm at the end of the word `return` in a JavaScript file.

I haven't narrowed down a reproduction yet, but I did find the line that is causing the error: https://github.com/minad/corfu/blob/cdc3e13ad312f5f12b3f78f842fff0b398eb4473/corfu.el#L686

For some reason, `beg` is `nil`

My completation-at-point-functions is set to `'(aj-capf-lsp)`, which is defined as:

```el
(defalias 'aj-capf-lsp (cape-capf-super #'lsp-completion-at-point #'ripgrep-capf))
```

(It's acutally defined slightly differently, but these two are the ones that are necessary to reproduce it.

`ripgrep-capf` is mine: https://github.com/aaronjensen/ripgrep-capf and it's certainly possible I'm doing something wrong there (or lsp-completion-at-point is). 

Any tips on diagnosing further? Thank you!

**Comment by minad** [comment]
Do you have a minimal recipe demonstrating the issue? From what I can tell, beg should never be nil in corfu--update, but maybe there is some code path I am missing.

**Reply by vifon** [reply]
I've just encountered this error after recently updating Corfu and having `corfu-echo-delay` set as `t` which is no longer supported (as of 60655764dc41dc42666abd69ff5c386661d98e59). I can reproduce it with the following minimal example and using the `M-tab` completion. Tested on cdc3e13, so currently `main~1`, but this new commit I just noticed shouldn't affect it.

```elisp
(require 'corfu)
(require 'corfu-echo)
(global-corfu-mode 1)
(corfu-echo-mode 1)
(setq corfu-echo-delay t)
```

In short, not a bug, just an outdated configuration.

**Reply by aaronjensen** [reply]
I don't have `corfu-echo-mode` enabled, nor `corfu-echo-delay` set, so my issue is not the same. I'll see what I can do to reproduce.

**Reply by vifon** [reply]
`corfu-popupinfo` is also affected, do you happen to have it enabled?

**Reply by aaronjensen** [reply]
@vifon I don't have that enabled. I did narrow it down to requiring corfu-precient.

@minad here is a repro: https://github.com/aaronjensen/corfu-repro

Open test.js, put the point after `return` and hit `tab`

**Reply by minad** [reply]
> I did narrow it down to requiring corfu-precient.

Okay, please report the issue there. As far as I know corfu-prescient advices Corfu at some places so something might go wrong there.

**Reply by aaronjensen** [reply]
Done, thanks. https://github.com/radian-software/prescient.el/issues/161

**Reply by aaronjensen** [reply]
I was able to reproduce it without prescient. Here is the trace:

```el
Debugger entered--Lisp error: (wrong-type-argument hash-table-p nil)
  lsp-completion--annotate("return")
  apply(lsp-completion--annotate "return" nil)
  #f(compiled-function (cand &rest args) #<bytecode -0xe3d70d047219c1b>)(#("return" 0 6 (face completions-common-part cape-capf-super ("return"))))
  corfu--affixate((#("return" 0 6 (face completions-common-part match-data (0 5 0 1 1 2 2 3 3 4 4 5) lsp-completion-prefix "retur" lsp-completion-markers (31 #<marker (moves after insertion) at 37 in test.js>) lsp-completion-start-point 31 lsp-sort-text "15" lsp-completion-item #<hash-table equal 6/24 0x580af067 ...>)) #("return" 0 6 (face completions-common-part cape-capf-super ("return")))))
  corfu--candidates-popup((#<window 3 on test.js> 31 (28 . 14) 0 nil 31 (4 . 1) nil (0 . 0) (7 . 14)))
  corfu--exhibit()
  #<subr corfu--post-command>()
  ad-Advice-corfu--post-command(#<subr corfu--post-command>)
  apply(ad-Advice-corfu--post-command #<subr corfu--post-command> nil)
  corfu--post-command()
```

It appears that `lsp-completion--annotate` is being invoked with completions from my other capf function. I don't know if that means my other capf function is misbehaving, or it means that `lsp-completion--annotate` only works with `lsp` completions. If that's the case, I may need to advise it to disregard plain strings.

Does this jive with what you would expect `cape-capf-super` to do (invoke annotation functions from one capf on another one's results?)

**Reply by minad** [reply, marked as answer]
Indeed. It could happen that annotation functions were called with candidates belonging to another Capf. Hopefully this commit fixes the problem:

https://github.com/minad/cape/commit/4fd90e3752aefe3ed38b0edc4b94214a5e64002f

**Reply by aaronjensen** [reply]
That appears to work. Thank you!

Somewhat related, is there a way to deduplicate the results between two capf when combining with `cape-capf-super`?

![CleanShot 2024-07-17 at 10 48 50@2x](https://github.com/user-attachments/assets/bb74bca6-9848-459a-bc85-cd079211d0df)

**Reply by minad** [reply]
Thanks for confirming. Deduplication only happens if candidates are equal also with respect to their text properties (`equal-including-properties`). This is intentional since equally named candidates may lead to different expansions, e.g., argument lists or snippets. https://github.com/minad/corfu/blob/5294a6ba0256230a7c92daf6b39afc351c45af9d/corfu.el#L602-L618
